internal/cache/async.go

```diff
@@ -0,0 +1,155 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+
+	"github.com/palantir/k8s-spark-scheduler/internal/cache/store"
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	rest "k8s.io/client-go/rest"
+)
+
+type object interface {
+	metav1.Object
+	runtime.Object
+}
+
+type asyncClient struct {
+	client             rest.Interface
+	resourceName       string
+	emptyObjectCreator func() object
+	queue              store.ShardedUniqueQueue
+	objectStore        store.ObjectStore
+}
+
+func (as *asyncClient) Run(ctx context.Context) {
+	for _, q := range as.queue.GetConsumers() {
+		go as.runWorker(ctx, q)
+	}
+}
+
+func (as *asyncClient) runWorker(ctx context.Context, requests <-chan func() store.Request) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case requestGetter := <-requests:
+			r := requestGetter()
+			switch r.Type {
+			case store.CreateRequestType:
+				as.doCreate(requestCtx(ctx, r.Key, "create"), r.Key)
+			case store.UpdateRequestType:
+				as.doUpdate(requestCtx(ctx, r.Key, "update"), r.Key)
+			case store.DeleteRequestType:
+				as.doDelete(requestCtx(ctx, r.Key, "delete"), r.Key)
+			}
+		}
+	}
+}
+
+func (as *asyncClient) doCreate(ctx context.Context, key store.Key) {
+	obj, ok := as.objectStore.Get(key)
+	if !ok {
+		svc1log.FromContext(ctx).Info("Ignoring request for deleted object")
+		return
+	}
+	result := as.emptyObjectCreator()
+	err := as.client.Post().
+		Namespace(obj.GetNamespace()).
+		Resource(as.resourceName).
+		Body(obj).
+		Do().
+		Into(result)
+	switch {
+	case err == nil:
+		as.objectStore.OverrideResourceVersionIfNewer(ctx, obj)
+	case isRetryableError(ctx, err):
+		as.queue.AddIfAbsent(store.CreateRequest(obj))
+	default:
+		logAndIgnore(ctx, err)
+	}
+}
+
+func (as *asyncClient) doUpdate(ctx context.Context, key store.Key) {
+	obj, ok := as.objectStore.Get(key)
+	if !ok {
+		svc1log.FromContext(ctx).Info("Ignoring request for deleted object")
+		return
+	}
+
+	result := as.emptyObjectCreator()
+	err := as.client.Put().
+		Namespace(obj.GetNamespace()).
+		Resource(as.resourceName).
+		Name(obj.GetName()).
+		Body(obj).
+		Do().
+		Into(result)
+	if err != nil {
+		as.objectStore.OverrideResourceVersionIfNewer(ctx, obj)
+	}
+	switch {
+	case err == nil:
+		as.objectStore.OverrideResourceVersionIfNewer(ctx, obj)
+	case isRetryableError(ctx, err):
+		as.queue.AddIfAbsent(store.UpdateRequest(obj))
+	case errors.IsConflict(err):
+		svc1log.FromContext(ctx).Warn("got conflict, will retry", svc1log.Stacktrace(err))
```

Comment:
How convinced are you that this is the right behaviour?

Reply:
Actually I am going to get the latest and update again here to force the update go through, I think retrying indefinitely is error prone